pkg/resourcemanager/controller/csrapprover/reconciler.go

```diff
@@ -166,3 +224,39 @@ func (r *Reconciler) mustApprove(ctx context.Context, csr *certificatesv1.Certif
 
 	return "all checks passed", true, nil
 }
+
+func (r *Reconciler) mustApproveKubeAPIServerClient(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) (string, decision, error) {
+	// Handle CSRs for gardener-node-agent only. There are other "kube-apiserver-client" CSRs which must not be touched.
+	if !strings.HasPrefix(x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix) {
+		return fmt.Sprintf("commonName %q is not prefixed with %q", x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix), csrNoOpinion, nil
+	}
+
+	machineName := strings.TrimPrefix(x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix)
+	machine := &machinev1alpha1.Machine{}
+	if err := r.SourceClient.Get(ctx, client.ObjectKey{Namespace: r.Config.MachineNamespace, Name: machineName}, machine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Sprintf("machine %q does not exist", machineName), csrDenied, nil
+		}
+		return "", csrNoOpinion, err
+	}
+
+	switch {
+	case strings.HasPrefix(csr.Spec.Username, "system:bootstrap:"):
+		if _, found := machine.Labels[machinev1alpha1.NodeLabelKey]; found {
+			return fmt.Sprintf("Cannot use bootstrap token since gardener-node-agent for machine %q is already bootstrapped", machineName), csrDenied, nil
+		}
+	case strings.HasPrefix(csr.Spec.Username, v1beta1constants.NodeAgentUserNamePrefix):
+		if csr.Spec.Username != x509cr.Subject.CommonName {
```

Comment:
You could add this condition to the line above to reduce nesting

Reply:
Then, I would have to add another empty `case` for the happy case. This would not be easier to understand, would it?

Reply:
OK, got it. Makes sense.